numba/targets/arraymath.py

```diff
@@ -3711,3 +3711,40 @@ def np_kaiser_impl(M, beta):
         return _i0n(n, alpha, beta)
 
     return np_kaiser_impl
+
+
+@overload(np.cross)
+def np_cross(a, b):
+    if isinstance(a, types.Array) and isinstance(b, types.Array):
+
+        dtype = np.promote_types(as_dtype(a.dtype), as_dtype(b.dtype))
+
+        def np_cross_impl(a, b):
+            if a.shape[-1] not in (2, 3) or b.shape[-1] not in (2, 3):
+                raise ValueError('dimension must be 2 or 3')
+
+
+            # Create the output array
+
+            if a.shape[-1] == 3 or b.shape[-1] == 3:
+                # Case: last output dim has 3 elements
+                if a.ndim == 1 and b.ndim == 1:
+                    # Case: inputs are 1D arrays
+                    cp = np.empty((3,), dtype)
+                else:
+                    # Case: inputs are ND arrays
+
+                    # we can't use np.broadcast; use np.add
+                    # since we only need the broadcast shape
+                    shape_ = np.add(a[..., 0], b[..., 0]).shape
```

Comment:
I am getting errors in this line when the example I'm testing (`a = [1, 2, 3]`, `b = [4, 5, 6]`) should enter the `a.ndim == 1 and b.ndim == 1` block.

Reply:
I tried this locally and get the following error:

```
Invalid use of Function(<function cross at 0x10eb59ea0>) with argument(s) of type(s): (array(int64, 1d, C), array(int64, 1d, C))
 * parameterized
In definition 0:
    TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Unknown attribute 'shape' of type int64

File "numba/targets/arraymath.py", line 3739:
        def np_cross_impl(a, b):
            <source elided>
                    # since we only need the broadcast shape
                    shape_ = np.add(a[..., 0], b[..., 0]).shape
                    ^

[1] During: typing of get attribute at /Users/vhaenel/git/numba/numba/targets/arraymath.py (3739)

File "numba/targets/arraymath.py", line 3739:
        def np_cross_impl(a, b):
            <source elided>
                    # since we only need the broadcast shape
                    shape_ = np.add(a[..., 0], b[..., 0]).shape
                    ^

    raised from /Users/vhaenel/git/numba/numba/typeinfer.py:915
In definition 1:
    TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Unknown attribute 'shape' of type int64

File "numba/targets/arraymath.py", line 3739:
        def np_cross_impl(a, b):
            <source elided>
                    # since we only need the broadcast shape
                    shape_ = np.add(a[..., 0], b[..., 0]).shape
                    ^

[1] During: typing of get attribute at /Users/vhaenel/git/numba/numba/targets/arraymath.py (3739)

File "numba/targets/arraymath.py", line 3739:
        def np_cross_impl(a, b):
            <source elided>
                    # since we only need the broadcast shape
                    shape_ = np.add(a[..., 0], b[..., 0]).shape
                    ^

    raised from /Users/vhaenel/git/numba/numba/typeinfer.py:915
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<function cross at 0x10eb59ea0>)
[2] During: typing of call at /Users/vhaenel/git/numba/numba/tests/test_np_functions.py (217)


File "numba/tests/test_np_functions.py", line 217:
def np_cross(a, b):
    return np.cross(a, b)
    ^
```

What I think is going on here is that the result of `np.add` becomes scalar and that doesn't have a `shape` attribute. Try commenting out the lines and at least your 1-d test should run.

Reply:
Also: the error is during just-in-time compilation of the function, not during the execution.